File: android/core/calc/src/test/kotlin/com/stocktracker/core/calc/TypeFilterTest.kt
package com.stocktracker.core.calc

import com.stocktracker.core.model.PositionType
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.Assert.assertEquals
import org.junit.Test
import java.io.File

/** Runs the same case table as src/utils/typeFilter.test.ts. */
class TypeFilterTest {
    private val cases = Json.parseToJsonElement(
        File(System.getProperty("user.dir")).resolve("../../../test-fixtures/type-filter/cases.json").readText(),
    ).jsonObject

    private fun types(o: JsonObject, key: String) =
        o.getValue(key).jsonArray.map { PositionType.valueOf(it.jsonPrimitive.content.uppercase()) }

    @Test fun effective() = cases.getValue("effective").jsonArray.forEach {
        val c = it.jsonObject
        assertEquals(c.toString(), types(c, "expected"), effectiveTypeFilter(types(c, "selected").toSet(), types(c, "held").toSet()).toList())
    }

    @Test fun toggle() = cases.getValue("toggle").jsonArray.forEach {
        val c = it.jsonObject
        val type = PositionType.valueOf(c.getValue("type").jsonPrimitive.content.uppercase())
        assertEquals(c.toString(), types(c, "expected"), toggleType(types(c, "selected").toSet(), type).toList())
    }

    @Test fun label() = cases.getValue("label").jsonArray.forEach {
        val c = it.jsonObject
        assertEquals(c.toString(), c.getValue("expected").jsonPrimitive.content, typeFilterLabel(types(c, "filter").toSet()))
    }
}

File: src/utils/typeFilter.test.ts
import { describe, expect, it } from 'vitest'
import cases from '../../test-fixtures/type-filter/cases.json'
import {
  effectiveTypeFilter, filterPositionsByType, parseStoredTypes, toggleType, typeFilterLabel, type AssetType,
} from './typeFilter'
import type { Position } from '../types'

describe('type filter (shared cases)', () => {
  it.each(cases.effective)('effective %j', ({ selected, held, expected }) => {
    expect([...effectiveTypeFilter(selected as AssetType[], held as AssetType[])]).toEqual(expected)
  })
  it.each(cases.toggle)('toggle %j', ({ selected, type, expected }) => {
    expect(toggleType(selected as AssetType[], type as AssetType)).toEqual(expected)
  })
  it.each(cases.label)('label %j', ({ filter, expected }) => {
    expect(typeFilterLabel(new Set(filter as AssetType[]))).toBe(expected)
  })
})

describe('type filter (web only)', () => {
  const p = (type: AssetType) => ({ id: type, ticker: type, type } as Position)
  it('an empty filter keeps every position', () => {
    expect(filterPositionsByType([p('stock'), p('etf')], new Set())).toHaveLength(2)
  })
  it('keeps only selected types', () => {
    expect(filterPositionsByType([p('stock'), p('etf')], new Set<AssetType>(['etf'])).map((x) => x.type)).toEqual(['etf'])
  })
  it('parses stored values defensively', () => {
    expect(parseStoredTypes(null)).toEqual([])
    expect(parseStoredTypes('not json')).toEqual([])
    expect(parseStoredTypes('["etf","bogus",3]')).toEqual(['etf'])
  })
})
